pkg/controller/atlascustomrole/atlascustomrole_controller.go

```diff
@@ -0,0 +1,245 @@
+		independentSyncPeriod:    independentSyncPeriod,
+	}
+}
+
+// +kubebuilder:rbac:groups=atlas.mongodb.com,resources=atlascustomroles,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=atlas.mongodb.com,resources=atlascustomroles/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",resources=secrets,verbs=create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=events,verbs=create;patch
+
+// +kubebuilder:rbac:groups=atlas.mongodb.com,namespace=default,resources=atlascustomroles,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=atlas.mongodb.com,namespace=default,resources=atlascustomroles/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",namespace=default,resources=secrets,verbs=create;update;patch;delete
+// +kubebuilder:rbac:groups="",namespace=default,resources=events,verbs=create;patch
+
+func (r *AtlasCustomRoleReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	atlasCustomRole := &akov2.AtlasCustomRole{}
+
+	err := r.Client.Get(ctx, req.NamespacedName, atlasCustomRole)
+	objectNotFound := err != nil && apiErrors.IsNotFound(err)
+	failedToRetrieve := err != nil && !objectNotFound
+
+	switch {
+	case failedToRetrieve:
+		return r.fail(req, err), nil
+	case objectNotFound:
+		return r.notFound(req), nil
+	}
+
+	if customresource.ReconciliationShouldBeSkipped(atlasCustomRole) {
+		return r.skip(), nil
+	}
+
+	r.Log.Infow("-> Starting AtlasCustomRole reconciliation", "spec", atlasCustomRole.Spec, "status",
+		atlasCustomRole.GetStatus())
+	conditions := akov2.InitCondition(atlasCustomRole, api.FalseCondition(api.ReadyType))
+	workflowCtx := workflow.NewContext(r.Log, conditions, ctx)
+	defer func() {
+		statushandler.Update(workflowCtx, r.Client, r.EventRecorder, atlasCustomRole)
+		r.Log.Infow("-> Finished AtlasCustomRole reconciliation", "spec", atlasCustomRole.Spec, "status",
+			atlasCustomRole.GetStatus())
+	}()
+
+	valid, err := customresource.ResourceVersionIsValid(atlasCustomRole)
+	if err != nil {
+		return r.terminate(workflowCtx, atlasCustomRole, api.ResourceVersionStatus, workflow.AtlasResourceVersionIsInvalid, true, err), nil
+	}
+
+	if !valid {
+		return r.terminate(workflowCtx,
+			atlasCustomRole,
+			api.ResourceVersionStatus,
+			workflow.AtlasResourceVersionMismatch,
+			true,
+			fmt.Errorf("version of the resource '%s' is higher than the operator version '%s'", atlasCustomRole.GetName(), version.Version)), nil
+	}
+	workflowCtx.SetConditionTrue(api.ResourceVersionStatus).SetConditionTrue(api.ValidationSucceeded)
+
+	if !r.AtlasProvider.IsResourceSupported(atlasCustomRole) {
+		return r.terminate(workflowCtx, atlasCustomRole,
+			api.ProjectCustomRolesReadyType, workflow.AtlasGovUnsupported,
+			false,
+			fmt.Errorf("the %T is not supported by Atlas for government", atlasCustomRole)), nil
+	}
+
+	credentials, err := selectCredentials(workflowCtx.Context, r.Client, atlasCustomRole)
```

Comment:
We should also be consistent with how we solve the credentials and populate the project ID from external reference or Kubernetes. I believe that is s single operation which is always teh same for all CRDs and controllers that implement independent resource usage.

The current way this is solved can be seen in the database users controller:
https://github.com/mongodb/mongodb-atlas-kubernetes/blob/main/pkg/controller/atlasdatabaseuser/databaseuser.go#L48

And also in the deployment controller:
https://github.com/mongodb/mongodb-atlas-kubernetes/blob/main/pkg/controller/atlasdeployment/atlasdeployment_controller.go#L137

I am proposing a simplification that would allow use to just use the same code centrally with minimal rewrite on each new controller:
https://github.com/mongodb/mongodb-atlas-kubernetes/pull/1930/files#diff-d17d85797ec5b7b965ec7b3508aab8f74d8bc04e930828939a234525efcaaddcR108

This is aided by a reconciler base implementation that does the heavy lifting for all reconcilers using composition:
https://github.com/mongodb/mongodb-atlas-kubernetes/pull/1930/files#diff-1d07cdba1ef7647d2e1ac39e6f7a861bbb19e44824dafd0f069d0abe96856e4dR40

But while we decide whether we want to go that way or not, I would suggest we stick with the schema we already merged in version 2.5 for this.

Reply:
I find this one a bit misleading. The methods are called `getProject` but they do more than that. That's why I made an alternative

Reply:
To get the project ID you need the credentials. And you also need the credentials to populate the services. Maybe the naming is not the best, but we already had a way, and we should be consistent.

We can have competing alternatives, but we should converge to a single way in the end.

My proposal in draft solves consistency by having a single implementation all controllers who need it can just use. We can argue on naming and refactoring, but I would rather converge on a single way that having one way per controller.

BTW, in here there is no extraction of the project ID, is it? Can you elaborate on that?

Reply:
I created CLOUDP-284794 so we discuss design beyond this particular PR